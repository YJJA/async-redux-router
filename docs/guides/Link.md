# Link
  as same as React Router Link

  ```
    import {Link} from 'async-redux-router'

    <Link to="/">Home</Link>
  ```
