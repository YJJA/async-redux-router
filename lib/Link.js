'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var isModifiedEvent = function isModifiedEvent(event) {
  return !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey);
};

/**
 * The public API for rendering a history-aware <a>.
 */

var Link = function (_React$Component) {
  (0, _inherits3.default)(Link, _React$Component);

  function Link() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Link);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Link.__proto__ || (0, _getPrototypeOf2.default)(Link)).call.apply(_ref, [this].concat(args))), _this), _this.handleClick = function (event) {
      if (_this.props.onClick) {
        _this.props.onClick(event);
      }

      if (!event.defaultPrevented && // onClick prevented default
      event.button === 0 && // ignore right clicks
      !_this.props.target && // let browser handle "target=_blank" etc.
      !isModifiedEvent(event) // ignore clicks with modifier keys
      ) {
          event.preventDefault();

          var history = _this.context.router.history;
          var _this$props = _this.props,
              replace = _this$props.replace,
              to = _this$props.to;


          if (replace) {
            history.replace(to);
          } else {
            history.push(to);
          }
        }
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Link, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          replace = _props.replace,
          to = _props.to,
          props = (0, _objectWithoutProperties3.default)(_props, ['replace', 'to']); // eslint-disable-line no-unused-vars

      var href = this.context.router.history.createHref(typeof to === 'string' ? { pathname: to } : to);

      return _react2.default.createElement('a', (0, _extends3.default)({}, props, { onClick: this.handleClick, href: href }));
    }
  }]);
  return Link;
}(_react2.default.Component);

Link.propTypes = {
  onClick: _propTypes2.default.func,
  target: _propTypes2.default.string,
  replace: _propTypes2.default.bool,
  to: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.object]).isRequired
};
Link.defaultProps = {
  replace: false
};
Link.contextTypes = {
  router: _propTypes2.default.shape({
    history: _propTypes2.default.shape({
      push: _propTypes2.default.func.isRequired,
      replace: _propTypes2.default.func.isRequired,
      createHref: _propTypes2.default.func.isRequired
    }).isRequired
  }).isRequired
};
exports.default = Link;