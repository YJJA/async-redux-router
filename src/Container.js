import React, {Component} from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

class Container extends Component {
  static propTypes = {
    router: PropTypes.shape({
      Components: PropTypes.array,
      match: PropTypes.object,
      history: PropTypes.object.isRequired
    }).isRequired
  };

  static childContextTypes = {
    router: PropTypes.object.isRequired
  };

  getChildContext() {
    return {
      router: {
        history: this.props.router.history
      }
    }
  }
  render() {
    const {history, Components, match} = this.props.router

    // Recursive rendering
    const renderComponents = (Components) => {
      let [Component, ...Children] = Components
      let children = null

      if (!Component) {
        return null
      }

      if (Children.length) {
        children = renderComponents(Children)
      }
      if (children) {
        return <Component history={history} match={match}>{children}</Component>
      }

      return <Component history={history} match={match} />
    }

    return renderComponents(Components)
  }
}

function mapStateToProps({router}) {
  return {router}
}

export default connect(
  mapStateToProps
)(Container)
