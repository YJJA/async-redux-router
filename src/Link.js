// https://github.com/ReactTraining/react-router/blob/master/packages/react-router-dom/modules/Link.js

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

const isModifiedEvent = (event) =>
  !!(event.metaKey || event.altKey || event.ctrlKey || event.shiftKey)

/**
 * The public API for rendering a history-aware <a>.
 */
class Link extends React.Component {
  static propTypes = {
    router: PropTypes.shape({
      history: PropTypes.shape({
        push: PropTypes.func.isRequired,
        replace: PropTypes.func.isRequired,
        createHref: PropTypes.func.isRequired
      }).isRequired
    }).isRequired,
    onClick: PropTypes.func,
    target: PropTypes.string,
    replace: PropTypes.bool,
    to: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object
    ]).isRequired
  };

  static defaultProps = {
    replace: false
  }

  handleClick = (event) => {
    if (this.props.onClick) {
      this.props.onClick(event)
    }

    if (
      !event.defaultPrevented && // onClick prevented default
      event.button === 0 && // ignore right clicks
      !this.props.target && // let browser handle "target=_blank" etc.
      !isModifiedEvent(event) // ignore clicks with modifier keys
    ) {
      event.preventDefault()

      const { history } = this.props.router
      const { replace, to } = this.props

      if (replace) {
        history.replace(to)
      } else {
        history.push(to)
      }
    }
  }

  render() {
    const { replace, to, router, dispatch, ...props } = this.props // eslint-disable-line no-unused-vars

    const href = router.history.createHref(
      typeof to === 'string' ? { pathname: to } : to
    )

    return <a {...props} onClick={this.handleClick} href={href}/>
  }
}

export default connect(
  ({router}) => ({router})
)(Link)
