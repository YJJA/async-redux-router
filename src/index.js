export {default as Container} from './Container'
export {default as Link} from './Link'
export {default as NavLink} from './NavLink'
export {default as routerReducer} from './routerReducer'
export {default as asyncContainer} from './asyncContainer'
