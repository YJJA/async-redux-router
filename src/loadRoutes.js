const loadRoutes = (store, location, query, branch) => {
  const promise = branch.map(({match, route}) => {
    if (typeof window !== 'object' && !route.server) {
      return null
    }
    if (route.component) {
      return route.component
    }
    return route.asyncComponent()
  })
  return Promise.all(promise)
    .then((Components) => {
      // Initializing Components data
      const promise = Components.map(Component => {
        if (Component && typeof Component.dispatchInitialAction === 'function') {
          return Component.dispatchInitialAction({store, location, query})
        }
        return null
      })

      const promiseAll = Promise.all(promise)
      if (typeof window === 'undefined') {
        return promiseAll.then(() => Components)
      } else {
        promiseAll.catch(err => console.log(err))
        return Components
      }
    })
}

export default loadRoutes
