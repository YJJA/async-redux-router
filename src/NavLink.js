import React, {Component} from 'react'
import PropTypes from 'prop-types'
import Link from './Link'
import matchPath from './matchPath'
import {connect} from 'react-redux'

class NavLink extends Component {
  static propTypes = {
    router: PropTypes.shape({
      history: PropTypes.object.isRequired
    }).isRequired,
    to: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object
    ]).isRequired,
    exact: PropTypes.bool,
    strict: PropTypes.bool,
    activeClass: PropTypes.string,
    style: PropTypes.object
  };

  static defaultProps = {
    activeClassName: 'active'
  };

  render() {
    const {location} = this.props.router.history
    const {activeClassName, className, to, exact, strict, ...props} = this.props
    const isActive = matchPath(location.pathname, { path: to, exact, strict })

    return <Link {...props} to={to} className={isActive ? [className, activeClassName].join(' ') : className} />
  }
}

export default connect(
  ({router}) => ({router})
)(NavLink)
