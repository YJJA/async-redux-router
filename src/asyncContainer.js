import _isString from 'lodash/isString'
import _isFunction from 'lodash/isFunction'
import _isPlainObject from 'lodash/isPlainObject'
import _isEmpty from 'lodash/isEmpty'

import loadRoutes from './loadRoutes'
import matchRoutes from './matchRoutes'
import routerChange from './routerActions'
import qs from './querystring'

let unlisten = null

const asyncContainer = (store, history, routes) => {
  const branch = matchRoutes(routes, history.location.pathname)
  const state = store.getState()
  const redirect = getRedirect(state, branch, history.location)

  if (redirect && redirect.to) {
    if (process.env.NODE_ENV !== 'production') {
      console.log(`No permission to access this route, redirect toï¼š${redirect.to}`)
    }
    // if server-side
    if (typeof window === 'undefined') {
      return Promise.resolve(redirect.to)
    }

    // other web site
    if (/^(https?):\/\//.test(redirect.to)) {
      window.location.href = redirect.to
      return Promise.resolve()
    }

    if (redirect.replace) {
      history.replace(redirect.to)
    } else {
      history.push(redirect.to)
    }

    return asyncContainer(store, history, routes)
  }

  if (!unlisten) {
    unlisten = history.listen(() => asyncContainer(store, history, routes))
  }

  const match = branch[branch.length - 1].match
  const query = qs.parse(history.location.search.substr(1))
  return loadRoutes(store, history.location, query, branch)
    .then(Components => {
      store.dispatch(routerChange(history, Components, {...match, query}))
    })
}

export default asyncContainer

function getRedirect(state, branch, from_uri) {
  let redirect = {
    to: '',
    replace: true
  }
  let route, auth, to, i
  for (i = 0; i < branch.length; i++) {
    route = branch[i].route
    auth = route.redirect || route.auth
    if (_isString(auth)) {
      to = auth
    } else if (_isFunction(auth)) {
      to = auth(state, history.location)
    }

    if (!_isEmpty(to)) {
      if (_isString(to)) {
        return {...redirect, to}
      } else if(_isPlainObject(to)) {
        return {...redirect, ...to}
      }
    }
  }
}
