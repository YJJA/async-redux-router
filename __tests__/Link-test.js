import React from 'react'
import {Link} from '../src/Link'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'

test('Link create has success', () => {
  const mockProps = {
    router: {
      history: {
        push: jest.fn(),
        replace: jest.fn(),
        createHref: jest.fn()
      }
    },
    onClick: jest.fn(),
    to: 'http://facebook.github.io'
  }
  const link = renderer.create(
    <Link {...mockProps}>Facebook Github</Link>
  )

  let tree = link.toJSON()
  expect(tree).toMatchSnapshot()
})

test('Link text check', () => {
  const mockProps = {
    router: {
      history: {
        push: jest.fn(),
        replace: jest.fn(),
        createHref: jest.fn()
      }
    },
    to: 'http://facebook.github.io'
  }

  const link = shallow(
    <Link {...mockProps}>Facebook Github</Link>
  )

  expect(link.text()).toEqual('Facebook Github')
})
