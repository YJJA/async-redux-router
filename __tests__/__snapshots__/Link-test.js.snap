// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link create has success 1`] = `
<a
  href={undefined}
  onClick={[Function]}
>
  Facebook Github
</a>
`;
